### PR TITLE
ui: peering

### PR DESCRIPTION
**Pull request by LevelbossMike**
### Description
Adds the MVP Ui for the peering feature.

### Testing & Reproduction steps
* Visit `http://localhost:4200/ui/<dc-name>/peers` and look at the UI - fixes https://github.com/hashicorp/consul/issues/13268
Screenshot 2022-06-22 at 13 59 14

* Services listing will include peered services - fixes https://github.com/hashicorp/consul/issues/13272
Screenshot 2022-06-22 at 14 00 26

* Nodes will include peered nodes
Screenshot 2022-06-22 at 14 05 07

* Topology view shows peering information -  fixes https://github.com/hashicorp/consul/issues/13273
Screenshot 2022-06-22 at 14 12 37

* Intentions will show peering information - fixes https://github.com/hashicorp/consul/issues/13274
Screenshot 2022-06-22 at 14 18 01

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [ ] not a security concern

WIP

cc @johncowen @evrowe

### CODE REVIEW

**Comment by Amier3** [comment]
Hey @LevelbossMike 

Glad to see another PR from you! Just a reminder to fill out the PR template when you get a chance 😄... so we can get this in the queue as soon as it's ready for review.

**Review by johncowen — COMMENTED** [review]
Hey @LevelbossMike, I gave it a quick spin and it looks great! (I know you said there was a little more to do still)

I scanned through the code also and I left a few comments there that are hopefully helpful, and it also helped me to formulate a question for something Evan and I were discussing last week so thanks for that!

Just so you know I might come back with a little more once I've played a little more with it, but if you could take a look at the queries here and let me know what you think in the meantime that would be great!

Thanks again!

P.S. @evrowe could you come back on the documentation links answer that Mike had in the code comment, I know we usually just leave these blank until available but maybe we have URLs for these already, ty!

File: ui/packages/consul-ui/app/templates/dc/services/instance.hbs [hunk 1]
```diff
@@ -151,6 +153,12 @@ as |item|}}
               <dt>Node Name</dt>
               <dd><a href="{{href-to 'dc.nodes.show' item.Node.Node}}">{{item.Node.Node}}</a></dd>
             </dl>
+            {{#if item.Service.PeerName}}
+              <dl>
+                <dt>Peer Name</dt>
+                <dd>{{item.Service.PeerName}}</dd>
+              </dl>
+            {{/if}}
```
[the remark this answers is not in the document]
Comment (LevelbossMike) [reply ui/packages/consul-ui/app/templates/dc/services/instance.hbs:161]: yes at some point but there's no peer detail route atm

File: ui/packages/consul-ui/app/models/peer.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import Model, { attr } from '@ember-data/model';
+
+export default class Peer extends Model {
+  @attr('string') name;
+  @attr('string') state;
+  @attr('string') createIndex;
+  @attr('string') modifyIndex;
+}
```
[the remark this answers is not in the document]
Comment (LevelbossMike) [reply ui/packages/consul-ui/app/models/peer.js:8]: ✅

**Review by johncowen — COMMENTED** [review]
Hey, sorry for the multiple review comment, just I'm finding things out as I go here and trying to come back to you as soon as I can.

I got info on the `with-imports` query param, and I added some inline comment in here in regards to that, let me know if thats not clear and I can explain more.

Apart from that, ideally it would be good if we did this without bringing tailwind in. We haven't used tailwind anywhere else in the application and we have remnants of 3 or 4 CSS framework/design system things in our application, and moving forwards there are no plans to migrate to use tailwind, so I'd rather not add an additional thing.

I'm pretty sure we can achieve the visuals here with styling that already exists in the application, but let me know either here or offline if you have any questions around that.

File: ui/packages/consul-ui/app/components/consul/service/list/index.hbs [hunk 3]
```diff
@@ -52,6 +55,7 @@
     <span>
       {{format-number item.InstanceCount}} {{pluralize item.InstanceCount 'instance' without-count=true}}
     </span>
+    <Consul::Peer @item={{item}} />
```
[the remark this answers is not in the document]
Comment (LevelbossMike) [reply ui/packages/consul-ui/app/components/consul/service/list/index.hbs:58]: This is located in another place than the "bucket-list" - i.e. in the header section. I'll look to pull in peer info into the bucket-list though where possible.

File: ui/packages/consul-ui/app/templates/dc/peers/index.hbs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+          <p>{{item.name}}</p>
+        </BlockSlot>
+        <BlockSlot @name="details">
+          <div class="flex items-center overflow-x-scroll space-x-4">
+            <Peerings::Badge @peering={{item}} />
+          </div>
+        </BlockSlot>
+      </ListCollection>
+      {{else}}
+      {{!-- TODO: do we need to check permissions here or will we receive an error automatically? --}}
```
[the remark this answers is not in the document]
Comment (LevelbossMike) [reply ui/packages/consul-ui/app/templates/dc/peers/index.hbs:24]: I'll add the permission handling when I finish the PR - This is still pending backend changes as far as I can tell. As soon as that is ready I'll add permission handling.

File: ui/packages/consul-ui/app/templates/dc/peers/index.hbs [hunk 5]
```diff
@@ -0,0 +1,55 @@
+<Route @name={{route}} as |route|>
+  <AppView>
+    <BlockSlot @name="header">
+      <h1>
+        <route.Title @title="Peers" />
+      </h1>
+    </BlockSlot>
+    <BlockSlot @name="toolbar">
+      <Peerings::Search @search={{this.filter}} @onSearch={{this.handleSearchChanged}} />
```
Comment (johncowen) [inline ui/packages/consul-ui/app/templates/dc/peers/index.hbs:9]: We have a search component that we use for all our searching, which I think I'll need to replace this with once this is handed over. So if you wanted to copy in the partition one in here and rename then great, but thats on my list for things to do once this PR is in, so just a heads up that I'll probably replace this with our component once this is merged in.

I'm kinda looking for the comment you had on the Firefox issue which is something you could help me with also though, I'll comment elsewhere once I see that.

File: ui/packages/consul-ui/app/components/peerings/badge/index.scss [hunk 6]
```diff
@@ -0,0 +1,42 @@
+.peerings-badge {
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  padding: 2px 8px;
+  border-radius: 5px;
+  gap: 4px;
```
[the remark this answers is not in the document]
Comment (LevelbossMike) [reply ui/packages/consul-ui/app/components/peerings/badge/index.scss:7]: [gap is only not supported by ie](https://caniuse.com/?search=gap) which has reached its [end of life](https://blogs.windows.com/windowsexperience/2022/06/15/internet-explorer-11-has-retired-and-is-officially-out-of-support-what-you-need-to-know/). what is the browser support matrix we are using at the moment?

File: ui/packages/consul-ui/app/components/peerings/badge/index.scss [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    color: #725a01;
+  }
+  &.terminated {
+    background: #f1f2f3;
+    color: #3b3d45;
+  }
+  &.undefined {
+    background: #f1f2f3;
+    color: #3b3d45;
+  }
```
[the remark this answers is not in the document]
Comment (LevelbossMike) [reply ui/packages/consul-ui/app/components/peerings/badge/index.scss:36]: I've taken these values out of the Figma designs. Does it make sense to keep these colors as is - that way we can switch them to the design system easily when we finally do that?

File: ui/packages/consul-ui/app/components/peerings/badge/index.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (johncowen) [reply ui/packages/consul-ui/app/components/peerings/badge/index.scss:7]: Oh yeah weird, I'd read this elsewhere, please ignore for the moment! I'll try and remember where I read this to confirm what I'd misunderstood

FYI: Browser support is here

https://github.com/hashicorp/consul/blob/main/ui/packages/consul-ui/config/targets.js#L28

File: ui/packages/consul-ui/app/components/peerings/badge/index.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (jnwright) [reply ui/packages/consul-ui/app/components/peerings/badge/index.scss:36]: Yeah, let's use the existing colors for now! 👍

File: ui/packages/consul-ui/app/templates/dc/peers/index.hbs [hunk 8]
```diff
@@ -0,0 +1,59 @@
+<Route @name={{route}} as |route|>
+  <Watcher @watch={{this.model.loadPeers}} as |w|>
+    {{did-insert w.fns.start}}
+    {{will-destroy w.fns.stop}}
+  </Watcher>
```
Comment (johncowen) [inline ui/packages/consul-ui/app/templates/dc/peers/index.hbs:5]: Are disconnection states something you are thinking of adding here? We have an idea of "disconnection" being the state after at least one successful HTTP request, but then erroring on a subsequent one. When this happens we show a notification at the top of the UI and then keep the last set of loaded data on the page so the user can still see the last thing they loaded.

Comment (LevelbossMike) [reply ui/packages/consul-ui/app/templates/dc/peers/index.hbs:5]: 👷‍♂️ That's why I was saying this component would ideally use a statechart to model its behavior - additional requirements tend to creep in into components like this and the component already is implementing rather intricate behavior. I don't know enough about the requirements you outlined in your comment so I suggest to add this as a follow up. Sounds like a useful addition though and it definitely sounds like something where requirements should be fleshed out explicitly before jumping into implementation.

File: ui/packages/consul-ui/app/templates/dc/peers/index.hbs (hunk 5 above)
Comment (LevelbossMike) [reply ui/packages/consul-ui/app/templates/dc/peers/index.hbs:13]: The existing component is pretty tightly coupled to how the custom router dsl sets up its search-properties. The peerings search isn't filterable by anything yet but will only search by name. That's why I decided to go with a custom implementation. Not sure if the crud stuff you are working on needs a new search component or adds additional search capabilities but feel free to change whatever you feel is necessary.

File: ui/yarn.lock [hunk 9]
```diff
@@ -2944,6 +2944,13 @@ ansi-styles@~1.0.0:
   resolved "https://registry.yarnpkg.com/ansi-styles/-/ansi-styles-1.0.0.tgz#cb102df1c56f5123eab8b67cd7b98027a0279178"
   integrity sha1-yxAt8cVvUSPquLZ817mAJ6AnkXg=
 
+ansi-to-html@^0.6.15:
+  version "0.6.15"
+  resolved "https://registry.yarnpkg.com/ansi-to-html/-/ansi-to-html-0.6.15.tgz#ac6ad4798a00f6aa045535d7f6a9cb9294eebea7"
+  integrity sha512-28ijx2aHJGdzbs+O5SNQF65r6rrKYnkuwTYm8lZlChuoJ9P1vVzIpWO20sQTqTPDXYp6NFwk326vApTtLVFXpQ==
+  dependencies:
+    entities "^2.0.0"
+
```
[the remark this answers is not in the document]
Comment (LevelbossMike) [reply ui/yarn.lock:2953]: seems to be pulled in by the ember-keyboard change

File: ui/packages/consul-ui/app/components/peerings/badge/index.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (jnwright) [reply ui/packages/consul-ui/app/components/peerings/badge/index.scss:36]: Sorry, I've [tweaked the colors](https://www.figma.com/file/xFhxiz7QKwit2q2EJQZ0JQ/Consul-Design-Library?node-id=3885%3A4547) a bit to make them more accessible. They're still using colors that already exist in the code

File: ui/packages/consul-ui/app/locations/fsm-with-optional.js [hunk 10]
```diff
@@ -196,6 +198,10 @@ export default class FSMWithOptionalLocation {
     if (typeof hash.partition !== 'undefined') {
       hash.partition = `_${hash.partition}`;
     }
+    if (typeof hash.peer !== 'undefined') {
+      hash.peer = `:${hash.peer}`;
+    }
+
```
[the remark this answers is not in the document]
Comment (johncowen) [reply ui/packages/consul-ui/app/locations/fsm-with-optional.js:204]: @LevelbossMike I think this ^ still needs looking at

**Pull request merged**